### PR TITLE
move rust crates to chromium_crate_io

### PR DESCRIPTION
**Pull request by bridiver**
Resolves https://github.com/brave/brave-browser/issues/36271

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by rillian** [comment]
As usual, it would help future archaeologists if you included an overview of the change and its motivation in the commit message and/or PR desecription. From a brief look at the code it seems like:

- Upstream chromium has changed the way they're vendoring third-party crates to use more of cargo's capabilities.
- Instead of `third_party/rust/crate_name/vX_Y/crate` the third party source is in `third_party/rust/chromium_crates_io/vendor/crate-X.Y.Z`. First party metadata and build description for `gn` remain in the old location with the third party source referenced from there.

The indent change on line 161 of script/brave_license_helper.py seems spurious (.vscode comment).

I noticed the adblock version numbers don't all match, but that's true of the main branch as well. Looks like things have been moved into `//brave/third_party/rust/chromium_crates_io/vendor/` rather than these being fresh copies from crates.io?

Splitting the source changes and file moves into separate commits might help the github review interface not fall over. :(

**Comment by bridiver** [comment]
> I noticed the adblock version numbers don't all match, but that's true of the main branch as well. Looks like things have been moved into //brave/third_party/rust/chromium_crates_io/vendor/ rather than these being fresh copies from crates.io?

This is definitely a fresh download from crates.io, although 3 days ago that may not have been the case

**Comment by bridiver** [comment]
> As usual, it would help future archaeologists if you included an overview of the change and its motivation in the commit message and/or PR desecription. From a brief look at the code it seems like:

I intend to squash the commits

File: script/brave_license_helper.py [hunk 1]
```diff
@@ -31,6 +31,11 @@ def AddBraveCredits(root, prune_paths, special_cases, prune_dirs,
         os.path.join('brave', 'vendor', 'brave-ios'),
         os.path.join('brave', 'vendor', 'brave_base'),
 
+        # These have auto-generated license files and
+        # GetThirdPartyDepsFromGNDepsOutput causes strange license errors
+        # unless the this entire directory is excluded.
+        os.path.join('brave', 'third_party', 'rust'),
```
Comment (rillian) [inline script/brave_license_helper.py:37]: I think this can just exclude `//brave/third_party/rust/chromium_crates_io/vendor`.

Comment (bridiver) [reply script/brave_license_helper.py:37]: please see the comment explaining why the entire directory is excluded

Comment (rillian) [reply script/brave_license_helper.py:37]: I read the comment, but I couldn't reproduce the errors. I suppose it doesn't matter much, but the narrower scope seemed a safer hack.

**Comment by rillian** [comment]
NB you can address the audit failure by applying [this patch](https://github.com/cloudflare/lol-html/pull/208) to `lol_html-0.3.1`.

File: script/brave_license_helper.py (hunk 1 above)
Comment (bridiver) [reply script/brave_license_helper.py:37]: I can try it again, but it failed with a very strange error before

File: third_party/rust/curve25519_dalek/v3/BUILD.gn [hunk 2]
```diff
@@ -1,97 +1,103 @@
...
     "//brave/third_party/rust/byteorder/v1:lib",
     "//brave/third_party/rust/digest/v0_9:lib",
     "//brave/third_party/rust/rand_core/v0_5:lib",
     "//brave/third_party/rust/subtle/v2:lib",
     "//brave/third_party/rust/zeroize/v1:lib",
-    "//third_party/rust/serde/v1:lib",
   ]
   features = [
     "alloc",
-    "serde",
```
Comment (bridiver) [inline third_party/rust/curve25519_dalek/v3/BUILD.gn:93]: double-checked this removal and it appears to be correct. This appears to only be enabled on v4

**Review by rillian — APPROVED** [review]
Works for me. We have some nits, and should figure out how to manage the `cargo vet` config four our own purposes, but I'd prefer to see that in a followup, since this will unblock other contributions involving rust code.

@fmarier are you ok with the license helper exemption until we can figure out how to tell it where the actual license files are?

**Review by fmarier — APPROVED** [review]
Licensing exception is okay, but it needs a follow-up issue to be filed so that we don't forget to figure out how to do this properly once the dust has settled.

**Comment by bridiver** [comment]
linux arm64 failure is unrelated to this PR. See https://github.com/brave/brave-core/pull/21443

File: patches/tools-crates-gnrt-gen.rs.patch [hunk 3]
```diff
@@ -0,0 +1,12 @@
+diff --git a/tools/crates/gnrt/gen.rs b/tools/crates/gnrt/gen.rs
```
Comment (bridiver) [inline patches/tools-crates-gnrt-gen.rs.patch:1]: at least some of these patches can go to upstream, we'll figure out the details in a follow-up

**Comment by bridiver** [comment]
There are no substantive changes here for p3a so I'm going to go ahead and merge

**Pull request merged**